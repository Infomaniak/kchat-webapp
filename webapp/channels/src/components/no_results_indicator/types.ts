// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum NoResultsVariant {
    Search = 'Search',
    Files = 'Files',
    ChannelSearch = 'ChannelSearch',
    Mentions = 'Mentions',
    FlaggedPosts = 'FlaggedPosts',
    PinnedPosts = 'PinnedPosts',
    ChannelFiles = 'ChannelFiles',
    ChannelFilesFiltered = 'ChannelFilesFiltered',
    UserGroups = 'UserGroups',
    UserGroupMembers = 'UserGroupMembers',
    UserGroupsArchived = 'UserGroupsArchived',
}

export enum NoResultsLayout {
    Vertical,
    Horizontal,
}
