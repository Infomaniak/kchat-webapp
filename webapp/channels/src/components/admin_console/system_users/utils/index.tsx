// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {SortingState} from '@tanstack/react-table';
import React from 'react';
import {FormattedMessage} from 'react-intl';

import {UserReportSortColumns, ReportSortDirection} from '@mattermost/types/reports';
import type {UserReportOptions, UserReport} from '@mattermost/types/reports';
import type {Team} from '@mattermost/types/teams';

import {PAGE_SIZES} from 'components/admin_console/list_table';

import type {AdminConsoleUserManagementTableProperties} from 'types/store/views';

import {ColumnNames, RoleFilters, StatusFilter, TeamFilters} from '../constants';
import type {TableOptions} from '../system_users';
import type {OptionType as TeamFilterOptionType} from '../system_users_filters_popover/system_users_filter_team';

export function convertTableOptionsToUserReportOptions(tableOptions?: TableOptions): UserReportOptions {
    return {
        page_size: tableOptions?.pageSize || PAGE_SIZES[0],
        from_column_value: tableOptions?.fromColumnValue,
        from_id: tableOptions?.fromId,
        direction: tableOptions?.direction,
        ...getSortColumnForOptions(tableOptions?.sortColumn),
        ...getSortDirectionForOptions(tableOptions?.sortIsDescending),
        ...getSearchFilterOption(tableOptions?.searchTerm),
        ...getTeamFilterOption(tableOptions?.filterTeam),
        ...getStatusFilterOption(tableOptions?.filterStatus),
        ...getRoleFilterOption(tableOptions?.filterRole),
        date_range: tableOptions?.dateRange,
    };
}

/**
 * Converts the sorting column name to API compatible sorting column name. Default sorting column name is by username.
 */
export function getSortColumnForOptions(id?: SortingState[0]['id']): Pick<UserReportOptions, 'sort_column'> {
    let sortColumn: UserReportOptions['sort_column'];

    if (id === ColumnNames.email) {
        sortColumn = UserReportSortColumns.email;
    } else if (id === ColumnNames.createAt) {
        sortColumn = UserReportSortColumns.createAt;
    } else {
        // Default sorting to first User details column
        sortColumn = UserReportSortColumns.username;
    }

    return {
        sort_column: sortColumn,
    };
}

/**
 * Converts the sorting direction to API compatible sorting direction. Default sorting direction is ascending.
 */
export function getSortDirectionForOptions(desc?: SortingState[0]['desc']): Pick<UserReportOptions, 'sort_direction'> {
    let sortDirection: UserReportOptions['sort_direction'];

    if (desc) {
        sortDirection = ReportSortDirection.descending;
    } else {
        sortDirection = ReportSortDirection.ascending;
    }

    return {
        sort_direction: sortDirection,
    };
}

/**
 * It returns the value of that column on which sorting is applied.
 */
export function getSortableColumnValueBySortColumn(row: UserReport, sortColumn: AdminConsoleUserManagementTableProperties['sortColumn']): string {
    switch (sortColumn) {
    case ColumnNames.email:
        return row.email;
    case ColumnNames.createAt:
        return String(row.create_at);
    default:
        return row.username;
    }
}

export function getTeamFilterOption(teamId?: string): Partial<Pick<UserReportOptions, 'team_filter' | 'has_no_team'>> {
    if (!teamId || teamId === TeamFilters.AllTeams) {
        return {
            team_filter: undefined,
            has_no_team: undefined,
        };
    } else if (teamId === TeamFilters.NoTeams) {
        return {
            team_filter: undefined,
            has_no_team: true,
        };
    }

    return {
        team_filter: teamId,
        has_no_team: undefined,
    };
}

export function getStatusFilterOption(status?: string): Partial<Pick<UserReportOptions, 'hide_active' | 'hide_inactive'>> {
    if (status === StatusFilter.Active) {
        return {
            hide_inactive: true,
        };
    } else if (status === StatusFilter.Deactivated) {
        return {
            hide_active: true,
        };
    }

    return {
        hide_active: undefined,
        hide_inactive: undefined,
    };
}

export function getPaginationInfo(pageIndex: number, pageSize: number, currentLength: number, total?: number) {
    if (!currentLength) {
        return (
            <FormattedMessage
                id='admin.system_users_list.pagination.no_users'
                defaultMessage='0 users'
            />
        );
    }

    const firstPage = (pageIndex * pageSize) + 1;
    const lastPage = (pageIndex * pageSize) + currentLength;
    const totalItems = total || 0;

    return (
        <FormattedMessage
            id='admin.system_users_list.pagination'
            defaultMessage='Showing {firstPage} - {lastPage} of {totalItems} users'
            values={{
                firstPage,
                lastPage,
                totalItems,
            }}
        />
    );
}

export function getDefaultSelectedValueFromList<T extends {value: string}>(value: string, options: T[]) {
    const option = options.find((option) => option.value === value);

    if (option) {
        return option;
    }

    return options[0];
}

export function getDefaultSelectedTeam(teamId: Team['id'] | string, label?: string): TeamFilterOptionType {
    if (!teamId || teamId === TeamFilters.AllTeams) {
        return {
            value: TeamFilters.AllTeams,
            label: (
                <FormattedMessage
                    id='admin.system_users.filters.team.allTeams'
                    defaultMessage='All Teams'
                />
            ),
        };
    } else if (teamId === TeamFilters.NoTeams) {
        return {
            value: TeamFilters.NoTeams,
            label: (
                <FormattedMessage
                    id='admin.system_users.filters.team.noTeams'
                    defaultMessage='No Teams'
                />
            ),
        };
    }

    return {
        value: teamId,
        label: label || '',
    };
}

export function getRoleFilterOption(role?: string): Pick<UserReportOptions, 'role_filter'> {
    if (!role || role === RoleFilters.Any) {
        return {role_filter: undefined};
    }
    return {role_filter: role};
}

export function getSearchFilterOption(search?: string): Pick<UserReportOptions, 'search_term'> {
    if (!search || search.trim().length === 0) {
        return {search_term: undefined};
    }

    return {search_term: search};
}
