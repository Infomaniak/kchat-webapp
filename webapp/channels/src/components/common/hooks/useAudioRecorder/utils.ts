export const getWasmFileURL = () =>
    new URL('wasm-media-encoders/wasm/mp3', import.meta.url);
