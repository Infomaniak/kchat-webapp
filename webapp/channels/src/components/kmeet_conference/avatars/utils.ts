export function countMeta<T>(
    items: T[],
    total = items.length,
): [T[], T[], {overflowUnnamedCount: number; nonDisplayCount: number}] {
    const breakAt = Math.max(items.length, total) > 4 ? 3 : 4;

    const displayItems = items.slice(0, breakAt);
    const overflowItems = items.slice(breakAt);

    const overflowUnnamedCount = Math.max(total - displayItems.length - overflowItems.length, 0);
    const nonDisplayCount = overflowItems.length + overflowUnnamedCount;

    return [displayItems, overflowItems, {overflowUnnamedCount, nonDisplayCount}];
}
