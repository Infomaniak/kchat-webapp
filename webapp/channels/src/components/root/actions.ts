// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {History} from 'history';

import type {UserProfile} from '@mattermost/types/users';

import {getFirstAdminSetupComplete} from 'mattermost-redux/actions/general';
import {getProfiles} from 'mattermost-redux/actions/users';
import {General} from 'mattermost-redux/constants';
import {getIsOnboardingFlowEnabled} from 'mattermost-redux/selectors/entities/preferences';
import {getActiveTeamsList} from 'mattermost-redux/selectors/entities/teams';
import {checkIsFirstAdmin, getCurrentUser, isCurrentUserSystemAdmin} from 'mattermost-redux/selectors/entities/users';
import type {ThunkActionFunc} from 'mattermost-redux/types/actions';

import * as GlobalActions from 'actions/global_actions';

import {StoragePrefixes} from 'utils/constants';

export function redirectToOnboardingOrDefaultTeam(history: History): ThunkActionFunc<void> {
    return async (dispatch, getState) => {
        const state = getState();
        const isUserAdmin = isCurrentUserSystemAdmin(state);
        if (!isUserAdmin) {
            GlobalActions.redirectUserToDefaultTeam();
            return;
        }

        const teams = getActiveTeamsList(state);

        const onboardingFlowEnabled = getIsOnboardingFlowEnabled(state);

        if (teams.length > 0 || !onboardingFlowEnabled) {
            GlobalActions.redirectUserToDefaultTeam();
            return;
        }

        const firstAdminSetupComplete = await dispatch(getFirstAdminSetupComplete());
        if (firstAdminSetupComplete?.data) {
            GlobalActions.redirectUserToDefaultTeam();
            return;
        }

        const profilesResult = await dispatch(getProfiles(0, General.PROFILE_CHUNK_SIZE, {roles: General.SYSTEM_ADMIN_ROLE}));
        if (profilesResult.error) {
            GlobalActions.redirectUserToDefaultTeam();
            return;
        }
        const currentUser = getCurrentUser(getState());
        const adminProfiles = profilesResult.data?.reduce(
            (acc: Record<string, UserProfile>, curr: UserProfile) => {
                acc[curr.id] = curr;
                return acc;
            },
            {},
        );
        if (adminProfiles && checkIsFirstAdmin(currentUser, adminProfiles)) {
            history.push('/preparing-workspace');
            return;
        }

        GlobalActions.redirectUserToDefaultTeam();
    };
}

export function handleLoginLogoutSignal(e: StorageEvent): ThunkActionFunc<void> {
    return (dispatch, getState) => {
        // when one tab on a browser logs out, it sets __logout__ in localStorage to trigger other tabs to log out
        const isNewLocalStorageEvent = (event: StorageEvent) => event.storageArea === localStorage && event.newValue;

        if (e.key === StoragePrefixes.LOGOUT && isNewLocalStorageEvent(e)) {
            console.log('detected logout from a different tab'); //eslint-disable-line no-console
            GlobalActions.emitUserLoggedOutEvent('/', false, false);
        }
        if (e.key === StoragePrefixes.LOGIN && isNewLocalStorageEvent(e)) {
            const isLoggedIn = getCurrentUser(getState());

            // make sure this is not the same tab which sent login signal
            // because another tabs will also send login signal after reloading
            if (isLoggedIn) {
                return;
            }

            // detected login from a different tab
            function reloadOnFocus() {
                location.reload();
            }
            window.addEventListener('focus', reloadOnFocus);
        }
    };
}
