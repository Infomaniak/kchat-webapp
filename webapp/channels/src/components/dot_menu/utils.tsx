// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type React from 'react';
import {EventTypes, TELEMETRY_CATEGORIES} from 'utils/constants';

import {trackEvent} from 'actions/telemetry_actions';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (e: ChangeEvent, suffix: string): void => {
    if (e.type === EventTypes.CLICK) {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.CLICK + '_' + suffix);
    } else {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.SHORTCUT + '_ ' + suffix);
    }
};
