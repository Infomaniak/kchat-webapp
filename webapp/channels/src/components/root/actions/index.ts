// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {History} from 'history';

import type {ServerError} from '@mattermost/types/errors';
import type {UserProfile} from '@mattermost/types/users';

import {GeneralTypes} from 'mattermost-redux/action_types';
import {logError} from 'mattermost-redux/actions/errors';
import {getClientConfig, getLicenseConfig, getFirstAdminSetupComplete} from 'mattermost-redux/actions/general';
import {redirectToErrorPageIfNecessary} from 'mattermost-redux/actions/helpers';
import {getMyPreferences} from 'mattermost-redux/actions/preferences';
import {getMyKSuites, getMyTeamMembers, getMyTeamUnreads} from 'mattermost-redux/actions/teams';
import {getMe, getProfiles} from 'mattermost-redux/actions/users';
import {Client4} from 'mattermost-redux/client';
import {General} from 'mattermost-redux/constants';
import {isCollapsedThreadsEnabled, getIsOnboardingFlowEnabled, getTeamsOrderPreference} from 'mattermost-redux/selectors/entities/preferences';
import {getActiveTeamsList, getTeams} from 'mattermost-redux/selectors/entities/teams';
import {checkIsFirstAdmin, getCurrentUser, isCurrentUserSystemAdmin} from 'mattermost-redux/selectors/entities/users';
import {getLastKSuiteSeenId} from 'mattermost-redux/utils/team_utils';

import {getMyMeets} from 'actions/calls';
import {redirectUserToDefaultTeam, emitUserLoggedOutEvent} from 'actions/global_actions';
import {updateTeamsOrderForUser} from 'actions/team_actions';

import {getHistory} from 'utils/browser_history';
import {ActionTypes, StoragePrefixes} from 'utils/constants';
import {isDesktopApp, isMacApp, isNotMacMas} from 'utils/user_agent';

import type {ActionFuncAsync, ThunkActionFunc} from 'types/store';
import type {Translations} from 'types/store/i18n';

export type TranslationPluginFunction = (locale: string) => Translations

/**
 * This function meant to be used in root.tsx component loads config, license and if user is logged in, it loads user and its related data.
 */
export function loadConfigAndMe(): ThunkActionFunc<Promise<{isLoaded: boolean; isMeRequested?: boolean}>> {
    return async (dispatch, getState) => {
        // attempt to load config and license regardless if user is logged in or not
        try {
            await Promise.all([
                dispatch(getClientConfig()),
                dispatch(getLicenseConfig()),
            ]);

            let redirect = true;
            if (!isMacApp() || !isNotMacMas()) {
                redirect = false;
            }

            if (redirect) {
                const forceMigrationError: ServerError = {
                    message: 'Maintenance mode',
                    status_code: 1,
                };
                redirectToErrorPageIfNecessary(forceMigrationError);
            }
        } catch (error) {
            dispatch(logError(error as ServerError));
            return {
                isLoaded: false,
            };
        }

        // Load user and its related data now that we know that user is logged in
        const serverVersion = getState().entities.general.serverVersion || Client4.getServerVersion();
        dispatch({type: GeneralTypes.RECEIVED_SERVER_VERSION, data: serverVersion});
        const kSuiteCall = await dispatch(getMyKSuites());
        const kSuites = getTeams(getState());

        const suiteArr = Object.values(kSuites);

        if (suiteArr.length > 0) {
            const lastKSuiteSeenId = getLastKSuiteSeenId();
            const sortedSuites = suiteArr.sort((a, b) => {
                if (a.id === lastKSuiteSeenId) {
                    return -1;
                }
                if (b.id === lastKSuiteSeenId) {
                    return 1;
                }
                return b.update_at - a.update_at;
            });
            const lastKSuiteSeen = sortedSuites[0];

            if (isDesktopApp()) {
                window.postMessage({
                    type: 'switch-server',
                    data: lastKSuiteSeen.display_name,
                }, window.origin);
            }

            try {
                await Promise.all([
                    dispatch(getMe()),
                    dispatch(getMyPreferences()),
                    dispatch(getMyTeamMembers()),
                    dispatch(getMyMeets()),
                ]);
                dispatch(getMyTeamUnreads(isCollapsedThreadsEnabled(getState())));

                if (kSuiteCall?.data && Array.isArray(kSuiteCall.data)) {
                    // IK change: The desktop logic for team order is based on the 'teams_order' preference.
                    // Therefore, we need to update the preference while respecting the order already established by the desktop.
                    // If there is a new team, we add it to the end of the array.
                    // If one of the teams in the preference has been deleted, we remove it.

                    const fetchedTeamIds = kSuiteCall.data.map((team) => team.id);
                    const state = getState();

                    // IK change: getTeamsOrderPreference returns an object, not an array
                    // We need to convert the CSV string in prefObj.value into a proper array
                    const prefObj = getTeamsOrderPreference(state);
                    const currentPreferences: string[] = prefObj?.value ? prefObj.value.split(',') : [];

                    // Filter out any preferences that no longer exist in fetchedTeamIds
                    const cleanedPreferences = currentPreferences.filter((id) => fetchedTeamIds.includes(id));

                    // Find any new teams that aren't already in the preferences
                    const missingTeamIds = fetchedTeamIds.filter((id) => !cleanedPreferences.includes(id));

                    // Combine the cleaned preferences with the missing teams appended at the end
                    const updatedPreferences = [...cleanedPreferences, ...missingTeamIds];

                    // Ensure uniqueness just in case
                    const uniquePreferences = [...new Set(updatedPreferences)];

                    // Check if the preferences actually changed
                    const hasChanged =
        uniquePreferences.length !== currentPreferences.length ||
        uniquePreferences.some((id, index) => id !== currentPreferences[index]);

                    // If there was a change, update the user's team order preferences
                    if (hasChanged) {
                        // eslint-disable-next-line no-console
                        console.log('[loadConfigAndMe] Syncing team order preferences after fetch:', uniquePreferences);
                        updateTeamsOrderForUser(uniquePreferences);
                    }
                }
            } catch (error) {
                dispatch(logError(error as ServerError));
                return {
                    isLoaded: false,
                };
            }
        } else if (!isDesktopApp()) {
            // we should not use getHistory in mattermost-redux since it is an import from outside the package, but what else can we do
            if (kSuiteCall && kSuiteCall.data) {
                getHistory().push('/error?type=no_ksuite');
            }
        }

        return {
            isLoaded: true,
            isMeRequested: true,
        };
    };
}

export function registerCustomPostRenderer(type: string, component: any, id: string): ActionFuncAsync {
    return async (dispatch) => {
        // piggyback on plugins state to register a custom post renderer
        dispatch({
            type: ActionTypes.RECEIVED_PLUGIN_POST_COMPONENT,
            data: {
                postTypeId: id,
                pluginId: id,
                type,
                component,
            },
        });
        return {data: true};
    };
}

export function redirectToOnboardingOrDefaultTeam(history: History, searchParams?: URLSearchParams): ThunkActionFunc<void> {
    return async (dispatch, getState) => {
        const state = getState();
        const isUserAdmin = isCurrentUserSystemAdmin(state);
        if (!isUserAdmin) {
            redirectUserToDefaultTeam(searchParams);
            return;
        }

        const teams = getActiveTeamsList(state);

        const onboardingFlowEnabled = getIsOnboardingFlowEnabled(state);

        if (teams.length > 0 || !onboardingFlowEnabled) {
            redirectUserToDefaultTeam(searchParams);
            return;
        }

        const firstAdminSetupComplete = await dispatch(getFirstAdminSetupComplete());
        if (firstAdminSetupComplete?.data) {
            redirectUserToDefaultTeam(searchParams);
            return;
        }

        const profilesResult = await dispatch(getProfiles(0, General.PROFILE_CHUNK_SIZE, {roles: General.SYSTEM_ADMIN_ROLE}));
        if (profilesResult.error) {
            redirectUserToDefaultTeam(searchParams);
            return;
        }
        const currentUser = getCurrentUser(getState());
        const adminProfiles = profilesResult.data?.reduce(
            (acc: Record<string, UserProfile>, curr: UserProfile) => {
                acc[curr.id] = curr;
                return acc;
            },
            {},
        );
        if (adminProfiles && checkIsFirstAdmin(currentUser, adminProfiles)) {
            history.push('/preparing-workspace');
            return;
        }

        redirectUserToDefaultTeam(searchParams);
    };
}

export function handleLoginLogoutSignal(e: StorageEvent): ThunkActionFunc<void> {
    return (dispatch, getState) => {
        // when one tab on a browser logs out, it sets __logout__ in localStorage to trigger other tabs to log out
        const isNewLocalStorageEvent = (event: StorageEvent) => event.storageArea === localStorage && event.newValue;

        if (e.key === StoragePrefixes.LOGOUT && isNewLocalStorageEvent(e)) {
            console.log('detected logout from a different tab'); //eslint-disable-line no-console
            emitUserLoggedOutEvent('/', false, false);
        }
        if (e.key === StoragePrefixes.LOGIN && isNewLocalStorageEvent(e)) {
            const isLoggedIn = getCurrentUser(getState());

            // make sure this is not the same tab which sent login signal
            // because another tabs will also send login signal after reloading
            if (isLoggedIn) {
                return;
            }

            // detected login from a different tab
            function reloadOnFocus() {
                location.reload();
            }
            window.addEventListener('focus', reloadOnFocus);
        }
    };
}
