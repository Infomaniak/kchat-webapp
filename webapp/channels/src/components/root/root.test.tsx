// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';
import type {RouteComponentProps} from 'react-router-dom';
import rudderAnalytics from 'rudder-sdk-js';

import {ServiceEnvironment} from '@mattermost/types/config';

import {GeneralTypes} from 'mattermost-redux/action_types';
import {Client4} from 'mattermost-redux/client';
import type {Theme} from 'mattermost-redux/selectors/entities/preferences';

import * as GlobalActions from 'actions/global_actions';
import store from 'stores/redux_store';

import Root from 'components/root/root';

import {StoragePrefixes} from 'utils/constants';

import type {ProductComponent} from 'types/store/plugins';

jest.mock('rudder-sdk-js', () => ({
    identify: jest.fn(),
    load: jest.fn(),
    page: jest.fn(),
    ready: jest.fn((callback) => callback()),
    track: jest.fn(),
}));

jest.mock('actions/telemetry_actions');

jest.mock('actions/global_actions', () => ({
    redirectUserToDefaultTeam: jest.fn(),
}));

jest.mock('utils/utils', () => {
    const original = jest.requireActual('utils/utils');

    return {
        ...original,
        localizeMessage: () => {},
        applyTheme: jest.fn(),
        makeIsEligibleForClick: jest.fn(),

    };
});

jest.mock('mattermost-redux/actions/general', () => ({
    setUrl: () => {},
}));

describe('components/Root', () => {
    const baseProps = {
        telemetryEnabled: true,
        telemetryId: '1234ab',
        noAccounts: false,
        showTermsOfService: false,
        theme: {} as Theme,
        actions: {
            loadConfigAndMe: jest.fn().mockImplementation(() => {
                return Promise.resolve({
                    data: false,
                });
            }),
            getFirstAdminSetupComplete: jest.fn(() => Promise.resolve({
                type: GeneralTypes.FIRST_ADMIN_COMPLETE_SETUP_RECEIVED,
                data: true,
            })),
            getProfiles: jest.fn(),
            migrateRecentEmojis: jest.fn(),
            savePreferences: jest.fn(),
            registerCustomPostRenderer: jest.fn(),
            initializeProducts: jest.fn(),
        },
        permalinkRedirectTeamName: 'myTeam',
        showLaunchingWorkspace: false,
        plugins: [],
        products: [],
        ...{
            location: {
                pathname: '/',
            },
        } as RouteComponentProps,
        isCloud: false,
        rhsIsExpanded: false,
        rhsIsOpen: false,
        shouldShowAppBar: false,
    };

    test('should load config and license on mount and redirect to sign-up page', () => {
        const props = {
            ...baseProps,
            noAccounts: true,
            history: {
                push: jest.fn(),
            } as unknown as RouteComponentProps['history'],
        };

        const wrapper = shallow(<Root {...props}/>);

        (wrapper.instance() as any).onConfigLoaded();
        expect(props.history.push).toHaveBeenCalledWith('/signup_user_complete');
        wrapper.unmount();
    });

    test('should load user, config, and license on mount and redirect to defaultTeam on success', (done) => {
        document.cookie = 'MMUSERID=userid';
        localStorage.setItem('was_logged_in', 'true');

        const props = {
            ...baseProps,
            actions: {
                ...baseProps.actions,
                loadConfigAndMe: jest.fn().mockImplementation(() => {
                    return Promise.resolve({data: true});
                }),
            },
        };

        // Mock the method by extending the class because we don't have a chance to do it before shallow mounts the component
        class MockedRoot extends Root {
            onConfigLoaded = jest.fn(() => {
                expect(this.onConfigLoaded).toHaveBeenCalledTimes(1);
                expect(GlobalActions.redirectUserToDefaultTeam).toHaveBeenCalledTimes(1);
                expect(props.actions.loadConfigAndMe).toHaveBeenCalledTimes(1);
                done();
            });
        }

        const wrapper = shallow(<MockedRoot {...props}/>);
        wrapper.unmount();
    });

    test('should load user, config, and license on mount and should not redirect to defaultTeam id pathname is not root', (done) => {
        document.cookie = 'MMUSERID=userid';
        localStorage.setItem('was_logged_in', 'true');

        const props = {
            ...baseProps,
            location: {
                pathname: '/admin_console',
            } as RouteComponentProps['location'],
            actions: {
                ...baseProps.actions,
                loadConfigAndMe: jest.fn().mockImplementation(() => {
                    return Promise.resolve({data: true});
                }),
            },
        };

        // Mock the method by extending the class because we don't have a chance to do it before shallow mounts the component
        class MockedRoot extends Root {
            onConfigLoaded = jest.fn(() => {
                expect(this.onConfigLoaded).toHaveBeenCalledTimes(1);
                expect(GlobalActions.redirectUserToDefaultTeam).not.toHaveBeenCalled();
                expect(props.actions.loadConfigAndMe).toHaveBeenCalledTimes(1);
                done();
            });
        }

        const wrapper = shallow(<MockedRoot {...props}/>);
        wrapper.unmount();
    });

    test('should call history on props change', () => {
        const props = {
            ...baseProps,
            noAccounts: false,
            history: {
                push: jest.fn(),
            } as unknown as RouteComponentProps['history'],
        };
        const wrapper = shallow(<Root {...props}/>);
        expect(props.history.push).not.toHaveBeenCalled();
        const props2 = {
            noAccounts: true,
        };
        wrapper.setProps(props2);
        expect(props.history.push).toHaveBeenLastCalledWith('/signup_user_complete');
        wrapper.unmount();
    });

    test('should reload on focus after getting signal login event from another tab', () => {
        Object.defineProperty(window.location, 'reload', {
            configurable: true,
            writable: true,
        });
        window.location.reload = jest.fn();
        const wrapper = shallow(<Root {...baseProps}/>);
        const loginSignal = new StorageEvent('storage', {
            key: StoragePrefixes.LOGIN,
            newValue: String(Math.random()),
            storageArea: localStorage,
        });

        window.dispatchEvent(loginSignal);
        window.dispatchEvent(new Event('focus'));
        expect(window.location.reload).toBeCalledTimes(1);
        wrapper.unmount();
    });

    describe('onConfigLoaded', () => {
        afterEach(() => {
            Client4.telemetryHandler = undefined;
        });

        test('should not set a TelemetryHandler when onConfigLoaded is called if Rudder is not configured', () => {
            store.dispatch({
                type: GeneralTypes.CLIENT_CONFIG_RECEIVED,
                data: {
                    ServiceEnvironment: ServiceEnvironment.DEV,
                },
            });

            const wrapper = shallow(<Root {...baseProps}/>);

            Client4.trackEvent('category', 'event');

            expect(Client4.telemetryHandler).not.toBeDefined();

            wrapper.unmount();
        });

        test('should set a TelemetryHandler when onConfigLoaded is called if Rudder is configured', () => {
            store.dispatch({
                type: GeneralTypes.CLIENT_CONFIG_RECEIVED,
                data: {
                    ServiceEnvironment: ServiceEnvironment.TEST,
                },
            });

            const wrapper = shallow(<Root {...baseProps}/>);

            (wrapper.instance() as any).onConfigLoaded();

            Client4.trackEvent('category', 'event');

            expect(Client4.telemetryHandler).toBeDefined();

            wrapper.unmount();
        });

        test('should not set a TelemetryHandler when onConfigLoaded is called but Rudder has been blocked', () => {
            (rudderAnalytics.ready as any).mockImplementation(() => {
                // Simulate an error occurring and the callback not getting called
            });

            store.dispatch({
                type: GeneralTypes.CLIENT_CONFIG_RECEIVED,
                data: {
                    ServiceEnvironment: ServiceEnvironment.TEST,
                },
            });

            const wrapper = shallow(<Root {...baseProps}/>);

            (wrapper.instance() as any).onConfigLoaded();

            Client4.trackEvent('category', 'event');

            expect(Client4.telemetryHandler).not.toBeDefined();

            wrapper.unmount();
        });
    });

    describe('Routes', () => {
        test('Should mount public product routes', () => {
            const mainComponent = () => (<p>{'TestMainComponent'}</p>);
            const publicComponent = () => (<p>{'TestPublicProduct'}</p>);

            const props = {
                ...baseProps,
                products: [{
                    id: 'productwithpublic',
                    baseURL: '/productwithpublic',
                    mainComponent,
                    publicComponent,
                } as unknown as ProductComponent,
                {
                    id: 'productwithoutpublic',
                    baseURL: '/productwithoutpublic',
                    mainComponent,
                    publicComponent: null,
                } as unknown as ProductComponent],
            };

            const wrapper = shallow(<Root {...props}/>);

            (wrapper.instance() as any).setState({configLoaded: true});
            expect(wrapper).toMatchSnapshot();
            wrapper.unmount();
        });
    });
});
