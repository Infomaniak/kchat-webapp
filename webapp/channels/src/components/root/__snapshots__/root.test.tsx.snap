// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Root Routes Should mount public product routes 1`] = `
<RootProvider>
  <Connect(MobileViewWatcher) />
  <LuxonController />
  <PerformanceReporterController />
  <Switch>
    <Route
      component={[Function]}
      path="/error"
    />
    <HFRoute
      component={[Function]}
      path="/login"
    />
    <HFRoute
      component={[Function]}
      path="/access_problem"
    />
    <HFTRoute
      component={[Function]}
      path="/reset_password"
    />
    <HFTRoute
      component={[Function]}
      path="/reset_password_complete"
    />
    <HFRoute
      component={[Function]}
      path="/signup_user_complete"
    />
    <HFRoute
      component={[Function]}
      path="/should_verify_email"
    />
    <HFRoute
      component={[Function]}
      path="/do_verify_email"
    />
    <HFTRoute
      component={[Function]}
      path="/claim"
    />
    <LoggedInRoute
      component={[Function]}
      path="/terms_of_service"
    />
    <Route
      component={[Function]}
      path="/landing"
    />
    <Route
      path="/admin_console"
    >
      <Switch>
        <LoggedInRoute
          component={[Function]}
          path="/admin_console"
          theme={Object {}}
        />
        <Connect(RootRedirect) />
      </Switch>
    </Route>
    <LoggedInHFTRoute
      component={[Function]}
      path="/select_team"
    />
    <LoggedInHFTRoute
      component={[Function]}
      path="/oauth/authorize"
    />
    <LoggedInHFTRoute
      component={[Function]}
      path="/create_team"
    />
    <LoggedInRoute
      component={[Function]}
      path="/mfa"
    />
    <LoggedInRoute
      component={[Function]}
      path="/preparing-workspace"
    />
    <Redirect
      from="/_redirect/integrations/:subpath*"
      to="/myTeam/integrations/:subpath*"
    />
    <Redirect
      from="/_redirect/pl/:postid"
      to="/myTeam/pl/:postid"
    />
    <CompassThemeProvider
      theme={Object {}}
    >
      <WindowSizeObserver />
      <Connect(ModalController) />
      <Connect(Component) />
      <Connect(injectIntl(SystemNotice)) />
      <GlobalHeader />
      <CloudEffectsWrapper />
      <withRouter(Connect(TeamSidebar)) />
      <div
        className="main-wrapper"
      >
        <Switch>
          <Route
            key="productwithpublic-public"
            path="/productwithpublic/public"
            render={[Function]}
          />
          <Route
            key="productwithpublic"
            path="/productwithpublic"
            render={[Function]}
          />
          <Route
            key="productwithoutpublic"
            path="/productwithoutpublic"
            render={[Function]}
          />
          <LoggedInRoute
            component={[Function]}
            path="/:team([a-z0-9\\\\-_]+)"
            theme={Object {}}
          />
          <Connect(RootRedirect) />
        </Switch>
        <withRouter(Connect(SidebarRight)) />
      </div>
      <Connect(Pluggable)
        pluggableName="Global"
      />
      <AppBar />
      <Connect(Component) />
    </CompassThemeProvider>
  </Switch>
</RootProvider>
`;
